### ISSUE TITLE
Unable to mount secretVolume and configMapVolumes simultaneously

### ISSUE THREAD

**Issue by ilhaan** [pingidentity/helm-charts#126]
I have the following in my custom `values.yaml` for pingfederate-admin: 

```
pingfederate-admin:
  enabled: true

  secretVolumes:
    pingfederate-license:
      items:
        license: /opt/in/instance/server/default/conf/pingfederate.lic

  configMapVolumes:
    pingfederate-props:
      items:
        pf-props: /opt/in/etc/pingfederate.properties 
```

I created the secret and  configmap using the following kubectl commands: 

```
kubectl create secret generic pingfederate-license --from-file=license=<path-to-license-file>
kubectl create configmap pingfederate-props --from-file=pf-props=<path-to-properties-file>
```

However, the helm chart seems to only use the `configMapVolume` if both volumes are defined as shown in my `values.yaml` config. When I remove the `configMapVolumes` block, and re-run helm, the license mount shows up. 

I even tried switching the order these are defined in `values.yaml` and saw the same behavior in chart versions 0.5.1 and 0.5.3. 

Let me know if I am doing something wrong here and how I can fix this.

**Comment by ilhaan** [comment]
Adding `helm template` outputs to show the above described behavior. I used the `ping-devops-0.5.3` tagged release of the repo for these. 

* Custom `values.yaml` with both `secretVolumes` and `configMapVolumes` defined for `pingfederate-admin`: 
```
➜  ping-devops git:(29821d0) ✗ cat values-both.yaml                                                                                              
pingfederate-admin:
  enabled: true

  secretVolumes:
    pingfederate-license:
      items:
        license: /opt/in/instance/server/default/conf/pingfederate.lic

  configMapVolumes:
    pingfederate-props:
      items:
        pf-props: /opt/in/etc/pingfederate.properties 

➜  ping-devops git:(29821d0) ✗ helm template -s templates/pingfederate-admin/workload.yaml . -f values-both.yaml | yq '.spec.template.spec.volumes'
[
  {
    "configMap": {
      "items": [
        {
          "key": "pf-props",
          "path": "pingfederate.properties"
        }
      ],
      "name": "pingfederate-props"
    },
    "name": "pingfederate-props"
  }
]
```

* Similar values file with only `configMapVolumes` defined: 
```
➜  ping-devops git:(29821d0) ✗ cat values-cm.yaml                                                                                                  
pingfederate-admin:
  enabled: true

  configMapVolumes:
    pingfederate-props:
      items:
        pf-props: /opt/in/etc/pingfederate.properties 

➜  ping-devops git:(29821d0) ✗ helm template -s templates/pingfederate-admin/workload.yaml . -f values-cm.yaml | yq '.spec.template.spec.volumes'
[
  {
    "configMap": {
      "items": [
        {
          "key": "pf-props",
          "path": "pingfederate.properties"
        }
      ],
      "name": "pingfederate-props"
    },
    "name": "pingfederate-props"
  }
]
```

* Values file with only `secretVolumes` defined: 
```
➜  ping-devops git:(29821d0) ✗ cat values-sec.yaml                                                                                               
pingfederate-admin:
  enabled: true

  secretVolumes:
    pingfederate-license:
      items:
        license: /opt/in/instance/server/default/conf/pingfederate.lic

➜  ping-devops git:(29821d0) ✗ helm template -s templates/pingfederate-admin/workload.yaml . -f values-sec.yaml | yq '.spec.template.spec.volumes'
[
  {
    "name": "pingfederate-license",
    "secret": {
      "items": [
        {
          "key": "license",
          "path": "pingfederate.lic"
        }
      ],
      "secretName": "pingfederate-license"
    }
  }
]
```

* As mentioned in my previous post, I tested the same reversing the order of `configMapVolumes` and `secretVolumes` in `values.yaml`. Output shown below: 
```
➜  ping-devops git:(29821d0) ✗ cat values-reverse.yaml 
pingfederate-admin:
  enabled: true

  configMapVolumes:
    pingfederate-props:
      items:
        pf-props: /opt/in/etc/pingfederate.properties 

  secretVolumes:
    pingfederate-license:
      items:
        license: /opt/in/instance/server/default/conf/pingfederate.lic

➜  ping-devops git:(29821d0) ✗ helm template -s templates/pingfederate-admin/workload.yaml . -f values-reverse.yaml | yq '.spec.template.spec.volumes' 
[
  {
    "configMap": {
      "items": [
        {
          "key": "pf-props",
          "path": "pingfederate.properties"
        }
      ],
      "name": "pingfederate-props"
    },
    "name": "pingfederate-props"
  }
]
```

Seems to me that there is an issue here: https://github.com/pingidentity/helm-charts/blob/29821d02b7024a79d2d64ed2a5d991bfdbef21c9/charts/ping-devops/templates/pinglib/_workload.tpl#L291-L312

**Comment by tsigle** [comment]
@ilhaan,  Thank you for the excellent posting of the issue.

Indeed, I've confirmed that this is a bug.  I'll get a fix and new release out in the next few hours.

Thanks!

tls

**Comment by tsigle** [comment]
The issue will be resolved by having the workload only issue one `volumes:` in the yaml, as volumes contains an array, rather than a map.  This caused only the last volumes: to be used which caused the issue observed.

Fix is being tested now, and a release should show up shortly.

**Comment by ilhaan** [comment]
@tsigle Thanks for the quick update! Your fix seems to have fixed `volumes`, but I'm seeing the same issue with `volumeMount`: 

* `volumes`: 
```
➜  ping-devops git:(3cfbc7d) ✗ cat values-both.yaml 
pingfederate-admin:
  enabled: true

  secretVolumes:
    pingfederate-license:
      items:
        license: /opt/in/instance/server/default/conf/pingfederate.lic

  configMapVolumes:
    pingfederate-props:
      items:
        pf-props: /opt/in/etc/pingfederate.properties 

➜  ping-devops git:(3cfbc7d) ✗ helm template -s templates/pingfederate-admin/workload.yaml . -f values-both.yaml | yq '.spec.template.spec.volumes'             
[
  {
    "name": "pingfederate-license",
    "secret": {
      "items": [
        {
          "key": "license",
          "path": "pingfederate.lic"
        }
      ],
      "secretName": "pingfederate-license"
    }
  },
  {
    "configMap": {
      "items": [
        {
          "key": "pf-props",
          "path": "pingfederate.properties"
        }
      ],
      "name": "pingfederate-props"
    },
    "name": "pingfederate-props"
  }
]
```

* But only one `volumeMount`: 
```
➜  ping-devops git:(3cfbc7d) ✗ helm template -s templates/pingfederate-admin/workload.yaml . -f values-both.yaml | yq '.spec.template.spec.containers[].volumeMounts'
[
  {
    "mountPath": "/opt/in/etc/pingfederate.properties",
    "name": "pingfederate-props",
    "readOnly": true,
    "subPath": "pingfederate.properties"
  }
]
```

**Comment by tsigle** [comment]
@ilhaan , Thanks for the comment back.

Yes, I did miss the volumeMounts fix.  Will get that fixed and push out a new release in next couple of hours.

Thanks again,

tls

**Comment by tsigle** [comment]
Fix for both volumes: and volumeMounts: have been released in version 0.5.5
